kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionBuilderImpl.java

```diff
@@ -0,0 +1,186 @@
+import static io.delta.kernel.internal.util.Preconditions.checkArgument;
+import static io.delta.kernel.internal.util.VectorUtils.stringArrayValue;
+import static io.delta.kernel.internal.util.VectorUtils.stringStringMapValue;
+
+public class TransactionBuilderImpl implements TransactionBuilder {
+    private static final Logger logger = LoggerFactory.getLogger(TransactionBuilderImpl.class);
+
+    private final long currentTimeMillis = System.currentTimeMillis();
+    private final TableImpl table;
+    private final String engineInfo;
+    private final Operation operation;
+    private Optional<StructType> schema = Optional.empty();
+    private Optional<List<String>> partitionColumns = Optional.empty();
+
+    public TransactionBuilderImpl(TableImpl table, String engineInfo, Operation operation) {
+        this.table = table;
+        this.engineInfo = engineInfo;
+        this.operation = operation;
+    }
+
+    @Override
+    public TransactionBuilder withSchema(Engine engine, StructType newSchema) {
+        this.schema = Optional.of(newSchema); // will be verified as part of the build() call
+        return this;
+    }
+
+    @Override
+    public TransactionBuilder withPartitionColumns(Engine engine, List<String> partitionColumns) {
+        if (!partitionColumns.isEmpty()) {
+            this.partitionColumns = Optional.of(partitionColumns);
+        }
+        return this;
+    }
+
+    @Override
+    public Transaction build(Engine engine) {
+        SnapshotImpl snapshot;
+        try {
+            snapshot = (SnapshotImpl) table.getLatestSnapshot(engine);
+        } catch (TableNotFoundException tblf) {
+            logger.info(
+                    "Table {} doesn't exist yet. Trying to create a new table.",
+                    table.getPath(engine));
+            schema.orElseThrow(() -> new IllegalArgumentException(
+                    "Table doesn't exist yet. Must provide a new schema."));
+            // Table doesn't exist yet. Create an initial snapshot with the new schema.
+            Metadata metadata = getInitialMetadata();
+            Protocol protocol = getInitialProtocol();
+            LogReplay logReplay = getEmptyLogReplay(engine, metadata, protocol);
+            snapshot = new InitialSnapshot(
+                    table.getDataPath(),
+                    logReplay,
+                    metadata,
+                    protocol);
+        }
+
+        boolean isNewTable = snapshot.getVersion(engine) < 0;
+        validate(snapshot, isNewTable);
+
+        return new TransactionImpl(
+                isNewTable,
+                table.getDataPath(),
+                table.getLogPath(),
+                snapshot,
+                engineInfo,
+                operation,
+                snapshot.getProtocol(),
+                snapshot.getMetadata());
+    }
+
+    /**
+     * Validate the given parameters for the transaction.
+     */
+    private void validate(SnapshotImpl snapshot, boolean isNewTable) {
```

Comment:
also validate that the table isn't ONLY partition columns, i.e. there must be at least 1 data column